src/regexset.rs

```diff
@@ -0,0 +1,832 @@
+                                *pattern_id,
+                                RegexSetMatch {
+                                    pattern_index: *pattern_id,
+                                    captures: captures.clone(),
+                                },
+                            ));
+                        }
+                    }
+                }
+            }
+
+            // Check hard patterns - search each if not cached
+            for (i, hard_pattern) in self.set.inner.hard_patterns.iter().enumerate() {
+                if self.hard_cache[i].is_none() {
+                    match self.search_hard_pattern(i, hard_pattern) {
+                        Ok(result) => {
+                            self.hard_cache[i] = result;
+                        }
+                        Err(e) => return Some(Err(e)),
+                    }
+                }
+
+                if let Some((ref range, ref captures)) = self.hard_cache[i] {
+                    if range.start >= self.current_pos {
+                        let pattern_id = hard_pattern.pattern_id;
+                        let key = (range.start, pattern_id);
+                        if earliest_match.is_none()
+                            || key
+                                < (
+                                    earliest_match.as_ref().unwrap().0,
+                                    earliest_match.as_ref().unwrap().1,
+                                )
+                        {
+                            earliest_match = Some((
+                                range.start,
+                                pattern_id,
+                                RegexSetMatch {
+                                    pattern_index: pattern_id,
+                                    captures: captures.clone(),
+                                },
+                            ));
+                        }
+                    }
+                }
+            }
+
+            match earliest_match {
+                Some((_, _, match_result)) => {
+                    // Advance position for next iteration
+                    let match_len = match_result.end() - match_result.start();
+                    self.current_pos = match_result.end() + if match_len == 0 { 1 } else { 0 };
+
+                    // Invalidate cache entries that are now behind us
+                    self.invalidate_cache_before(self.current_pos);
+
+                    return Some(Ok(match_result));
+                }
+                None => {
+                    // No matches found
+                    return None;
+                }
+            }
+        }
+    }
+}
+
+impl<'h> RegexSetMatches<'h> {
+    fn search_easy_patterns(&mut self) -> Result<()> {
+        let mut cache = Vec::new();
+
+        if let Some(ref easy_set) = self.set.inner.easy_patterns {
+            let input = RaInput::new(self.haystack).range(self.range.clone());
+
+            // Find all matches from the DFA
```

Comment:
Is this what the spec said to do? If so, I suggest we change the spec, because this seems inefficient. Syntax highlighting could change context after a match, for example after parsing a `fn name(` it would switch to a new RegexSet containing patterns for parsing function parameters until it reaches `)`. It would be best if we find the first match in range, and remember how far in the string it searched so we can resume on the next iteration. In general I thought the idea was to take advantage of the way regex-automata can calculate which patterns matched more cheaply than where the patterns matched, and use that to skip searching for positions of patterns which don't appear in the haystack in the given range. But maybe because all patterns are in the same dfa it doesn't work like that?